src/golang/lib/workflow/preview_cache/preview_cache.go

```diff
@@ -82,19 +77,53 @@ func (c *inMemoryPreviewCacheManagerImpl) Put(ctx context.Context, artifactSigna
 	return c.putMulti(ctx, []uuid.UUID{artifactSignature}, []*utils.ExecPaths{execPaths})
 }
 
+func castCachedValueToEntry(val interface{}) *Entry {
+	entry, ok := val.(Entry)
+	if !ok {
+		return nil
+	}
+	return &entry
+}
+
+// isEqual checks whether an execPath has the same data as a preview cache entry.
+func isEqual(execPaths *utils.ExecPaths, entry *Entry) bool {
+	return *execPaths == utils.ExecPaths{
+		ArtifactContentPath:  entry.ArtifactContentPath,
+		ArtifactMetadataPath: entry.ArtifactMetadataPath,
+		OpMetadataPath:       entry.OpMetadataPath,
+	}
+}
+
 func (c *inMemoryPreviewCacheManagerImpl) putMulti(ctx context.Context, artifactSignatures []uuid.UUID, execPathsList []*utils.ExecPaths) error {
 	for i, signatures := range artifactSignatures {
+
 		// If the entry already exists, delete the data it points to, since the entry will be overridden.
-		val, ok := c.cache.Peek(signatures)
-		if ok {
-			deleteDataForEntry(ctx, c.storageConfig, val)
+		var existingEntry *Entry
+		val, exists := c.cache.Peek(signatures)
+		if exists {
+			existingEntry = castCachedValueToEntry(val)
+			if existingEntry == nil {
+				return errors.New("Preview Artifact Cache is storing an unexpected data structure.")
+			}
+
+			// If the entry already exists and has the same data, short-circuit as there is no need
+			// to update anything.
+			if isEqual(execPathsList[i], existingEntry) {
+				return nil
+			}
 		}
 
 		c.cache.Add(signatures, Entry{
 			ArtifactContentPath:  execPathsList[i].ArtifactContentPath,
 			ArtifactMetadataPath: execPathsList[i].ArtifactMetadataPath,
 			OpMetadataPath:       execPathsList[i].OpMetadataPath,
 		})
+
+		// After adding the new value to the cache, delete the orphaned data if this overwrote an existing
+		// entry. This is best-effort.
+		if existingEntry != nil {
+			deleteDataForEntry(ctx, c.storageConfig, *existingEntry)
+		}
```

Comment:
Will we ever reach here?

Reply:
I don't think it will. Can drop a note here. But I do think it's totally reasonable to have this logic, since it means the cache doesn't rely on any assumptions about the cache user. Eg. if we change the cache entry to be mutable in the future, we won't have to remember to include this logic, since it's already here.

Reply:
Yeah let's drop a note here.

Reply:
actually on second thought, it looks better without this logic now that we error out if the entries are not equal. So its just less confusing if we remove it.